source/bluetooth/MicroBitPartialFlashingService.cpp

```diff
@@ -0,0 +1,322 @@
+           MicroBitEvent evt(MICROBIT_ID_PFLASH_NOTIFICATION, END_OF_TRANSMISSION ,CREATE_AND_FIRE);          break;
+        }
+        case MICROBIT_STATUS:
+        {
+          /*
+           * Return the version of the Partial Flashing Service and the current BLE mode (application / pairing)
+           */
+          uint8_t flashNotificationBuffer[] = {MICROBIT_STATUS, PARTIAL_FLASHING_VERSION, MicroBitBLEManager::manager->getBLEMode()};
+          ble.gattServer().notify(partialFlashCharacteristicHandle, (const uint8_t *)flashNotificationBuffer, sizeof(flashNotificationBuffer));
+          break;
+        }
+        case MICROBIT_RESET:
+        {
+          /*
+           * data[1] determines which mode to reset into: MICROBIT_BLE_MODE_PAIRING or MICROBIT_BLE_MODE_APPLICATION
+           */
+           switch(data[1]) {
+             case MICROBIT_BLE_MODE_PAIRING:
+             {
+               MicroBitEvent evt(MICROBIT_ID_PFLASH_NOTIFICATION, MICROBIT_RESET ,CREATE_AND_FIRE);
+               break;
+             }
+             case MICROBIT_BLE_MODE_APPLICATION:
+             {
+               microbit_reset();
+               break;
+             }
+           }
+           break;
+        }
+    }
+  }
+}
+
+
+
+/**
+  * @param data - A pointer to the data to process
+  *
+  */
+void MicroBitPartialFlashingService::flashData(uint8_t *data)
+{
+        // Receive 16 bytes per packet
+        // Buffer 8 packets - 32 uint32_t // 128 bytes per block
+        // When buffer is full trigger partialFlashingEvent
+        // When write is complete notify app and repeat
+        // +-----------+---------+---------+----------+
+        // | 1 Byte    | 2 Bytes | 1 Byte  | 16 Bytes |
+        // +-----------+---------+---------+----------+
+        // | COMMAND   | OFFSET  | PACKET# | DATA     |
+        // +-----------+---------+---------+----------+
+        uint8_t packetNum = data[3];
+        /**
+          * Packets with packet num < packet count
+          * Ignore as part of retransmitted block
+          */
+        if(packetNum < packetCount)
+          return;
+
+        /**
+          * Check packet count
+          * If the packet count doesn't match send a notification to the client
+          * and set the packet count to the next block number
+          */
+        if(packetNum != ++packetCount)
+        {
+            uint8_t flashNotificationBuffer[] = {FLASH_DATA, 0xAA};
+            ble.gattServer().notify(partialFlashCharacteristicHandle, (const uint8_t *)flashNotificationBuffer, sizeof(flashNotificationBuffer));
+            packetCount = blockPacketCount + 3;
+            blockNum = 0;
+            return;
+        }
+
+        // Add to block
+        for(int x = 0; x < 4; x++)
+            block[(4*blockNum) + x] = data[(4*x) + 4] | data[(4*x) + 5] << 8 | data[(4*x) + 6] << 16 | data[(4*x) + 7] << 24;
+
+        // Actions
+        switch(blockNum) {
+            // blockNum is 0, set up offset
+            case 0:
```

Comment:
Can we update to the relevant #define command codes rather than literals please?